This Week in Glean: Bytes in Memory (on Android) (#13)
badboy/fnordig.de: _posts/2020-05-04-this-week-in-glean.md

---
permalink: "/{{ year }}/{{ month }}/{{ day }}/this-week-in-glean"
title: "This Week in Glean: Bytes in Memory (on Android)"
published_date: "2020-05-04 12:00:00 +0200"
layout: post.liquid
data:
  route: blog
  tags:
    - mozilla
    - rust
---

(“This Week in Glean” is a series of blog posts that the Glean Team at Mozilla is using to try to communicate better about our work. They could be release notes, documentation, hopes, dreams, or whatever: so long as it is inspired by Glean.)

Last week's blog post: [ This Week in Glean: Glean for Python on Windows](https://blog.mozilla.org/data/2020/04/27/this-week-in-glean-glean-for-python-on-windows/)
by Mike Droettboom.
All "This Week in Glean" blog posts are available on the [Mozilla Data blog](https://blog.mozilla.org/data/category/glean/).

---

> With the Glean SDK we follow in the footsteps of [other teams](https://github.com/mozilla/application-services) to build a cross-platform library to be used in mobile applications.

Reviewer: We're now beyond mobile, right? :-)

> In this blog post we're going to take a look at how we transport some rich data across the FFI boundary to be reused on the Kotlin side of things.

Reviewer: Can you provide a bit more context about what the article is about? By reading this I'd expect the article to talk about FFI stuff , which is something you mostly worked on. However, the article also describes bits of the upload mechanism, which is what most of the implementation is about. I'd spend a few more words on this. If this is about describing the upload mechanism as well, then I'd provide a super tl;dr about how it should work.

Author: ok with linking [to the docs?](https://github.com/mozilla/glean/blob/8e0599b4df6c1a08f985e4ad5328fcf81a56a084/docs/dev/core/internal/upload.md)

this is really more about the approach than about uploading.

Reviewer: > 
> 
> ok with linking [to the docs?](https://github.com/mozilla/glean/blob/8e0599b4df6c1a08f985e4ad5328fcf81a56a084/docs/dev/core/internal/upload.md)
Yes !

> this is really more about the approach than about uploading.

Perfect! Even just static exactly this somewhere in this paragraph would make the rest more coherent, I believe.

## The problem

The Glean SDK provides an [FFI API](https://github.com/mozilla/glean/tree/master/glean-core/ffi) that can be consumed by what we call language bindings.
For the most part we pass <abbr title="Plain Old Data">POD</abbr> over that API boundary:
integers of various sizes (a problem in and of itself if the sides disagree about certain integer sizes and signedness),
bools (but actually encoded as an 8-bit integer),
but also strings as pointers to null-terminated UTF-8 strings in memory (sometimes we encode data into JSON and pass that over as strings).

However for some internal mechanisms we needed to communicate a bit more data back and forth.
We wanted to have different _tasks_, where each task variant could have additional data.
Luckily this additional data is either some integers or a bunch of strings only and not further nested data.

So this is the data we have on the Rust side:

```rust
enum Task {
    Upload(Request),
    Wait,
    Done,
}

struct Request {
    id: String,
    url: String,
}
```

(This code is simplified for the sake of this blog post.

> You can find the full code online [in the Glean repository](https://github.com/mozilla/glean/blob/8e0599b4df6c1a08f985e4ad5328fcf81a56a084/glean-core/src/upload/mod.rs).)

Reviewer: ```suggestion
You can find the full code online [in the Glean repository](https://github.com/mozilla/glean/blob/8e0599b4df6c1a08f985e4ad5328fcf81a56a084/glean-core/src/upload/mod.rs#L38-L48).)
```

`getRustString` is a [small helper](https://github.com/mozilla/glean/blob/da1e014e378660f243b92ce6012b7b29188d72d8/glean-core/android/src/main/java/mozilla/telemetry/glean/rust/LibGleanFFI.kt#L40-L42)
to copy the null-terminated C-like string to a Kotlin string.

The FFI function on the Kotlin side is defined as:

```java
fun glean_get_next_task(): FfiTask.ByValue
```

The types we convert to and then work with in Kotlin are small classes around the data.
If there's no attached data it's an object.

```java
class Request(
    val id: String,
    val url: String,
) { }

sealed class Task {
    class Upload(val request: Request) : Task()

    object Wait : Task()

    object Done : Task()
}
```

As the final piece of this code on the Kotlin side we can now fetch new tasks, convert it to more convenient Kotlin objects and work with them:

```java
val incomingTask = LibGleanFFI.INSTANCE.glean_get_next_task()
when (val action = incomingTask.toTask()) {
    is Task.Upload -> upload(action.request.id, aciton.request.url)
    Task.Wait -> return Result.retry()
    Task.Done -> return Result.success()
}
```

## What's next?

> Currently this piece of code is under testing.

Reviewer: The upload parts are under testing, but the more general "passing data through FFI" part has been around for a long while. I'd recommend clarifying that a bit.

---
permalink: "/{{ year }}/{{ month }}/{{ day }}/this-week-in-glean"
title: "This Week in Glean: Bytes in Memory (on Android)"
published_date: "2020-05-04 12:00:00 +0200"
layout: post.liquid
data:
  route: blog
  tags:
    - mozilla
    - rust
---

(“This Week in Glean” is a series of blog posts that the Glean Team at Mozilla is using to try to communicate better about our work. They could be release notes, documentation, hopes, dreams, or whatever: so long as it is inspired by Glean.)

Last week's blog post: [ This Week in Glean: Glean for Python on Windows](https://blog.mozilla.org/data/2020/04/27/this-week-in-glean-glean-for-python-on-windows/)
by Mike Droettboom.

> All "This Week in Glean" blog posts are available on the [Mozilla Data blog](https://blog.mozilla.org/data/category/glean/).

Reviewer 2: I thought that the index in the Book of Glean was going to be the canonical index?

Author: Will change that

---

With the Glean SDK we follow in the footsteps of [other teams](https://github.com/mozilla/application-services) to build a cross-platform library to be used in both mobile and desktop applications alike.
In this blog post we're taking a look at how we transport some rich data across the FFI boundary to be reused on the Kotlin side of things.
We're using a recent example of a new API in Glean that will [drive the HTTP upload of pings](https://github.com/mozilla/glean/blob/8e0599b4df6c1a08f985e4ad5328fcf81a56a084/docs/dev/core/internal/upload.md),
but the concepts I'm explaining here apply more generally.

> *Note:* This blog post is not a good introduction on doing Rust on Android, but I do plan to write about that in the future as well.

Most of this implementation was done by [Bea](https://brizental.github.io/) and I have been merely a reviewer,
around for questions and currently responsible for running final tests on this feature.

## The problem

The Glean SDK provides an [FFI API](https://github.com/mozilla/glean/tree/master/glean-core/ffi) that can be consumed by what we call language bindings.

> For the most part we pass <abbr title="Plain Old Data">POD</abbr> over that API boundary:

Reviewer 2: Do `abbr` render properly on fnordig? On github's preview they're eaten.

Author: yes, they do!

integers of various sizes (a problem in and of itself if the sides disagree about certain integer sizes and signedness),

> bools (but actually encoded as an 8-bit integer),

Reviewer 2: I'd love a couple of words here as to why.

> but also strings as pointers to null-terminated UTF-8 strings in memory (sometimes we encode data into JSON and pass that over as strings).

Reviewer 2: ```suggestion
but also strings as pointers to null-terminated UTF-8 strings in memory (for test APIs we encode data into JSON and pass that over as strings).
```

            }
            TaskTag.Wait.ordinal -> Task.Wait
            else -> Task.Done
        }
    }
}
```

This also defines the conversion to a new Kotlin type that eases usage on the Kotlin side.
`getRustString` is a [small helper](https://github.com/mozilla/glean/blob/da1e014e378660f243b92ce6012b7b29188d72d8/glean-core/android/src/main/java/mozilla/telemetry/glean/rust/LibGleanFFI.kt#L40-L42)
to copy the null-terminated C-like string to a Kotlin string.

The FFI function on the Kotlin side is defined as:

```java
fun glean_get_next_task(): FfiTask.ByValue
```

The types we convert to and then work with in Kotlin are small classes around the data.
If there's no attached data it's an object.

```java
class Request(
    val id: String,
    val url: String,
) { }

sealed class Task {
    class Upload(val request: Request) : Task()

    object Wait : Task()

    object Done : Task()
}
```

As the final piece of this code on the Kotlin side we can now fetch new tasks, convert it to more convenient Kotlin objects and work with them:

```java
val incomingTask = LibGleanFFI.INSTANCE.glean_get_next_task()
when (val action = incomingTask.toTask()) {

>     is Task.Upload -> upload(action.request.id, aciton.request.url)

Reviewer 2: ```suggestion
    is Task.Upload -> upload(action.request.id, action.request.url)
```
